### PR TITLE
make it async with Promises

### PR DESCRIPTION
**Pull request by giladgray**
gotta love async/await

### CODE REVIEW

File: cli.js [hunk 1]
```diff
@@ -34,10 +34,11 @@ const plugins = [].concat(argv.use)
...
-const doc = Documentalist.create();
+const docs = Documentalist.create();
 plugins.forEach(({ pattern, plugin }) => {
-    doc.use(new RegExp(pattern), plugin);
+    docs.use(new RegExp(pattern), plugin);
 });
-const documentation = doc.documentGlobs(...argv._);
+docs.documentGlobs(...argv._)
+    .then((data) => JSON.stringify(data, null, 2))
+    .then(console.log);
```
Comment (themadcreator) [inline cli.js]: 👏

File: src/documentalist.ts [hunk 2]
```diff
@@ -140,24 +140,23 @@ export class Documentalist<T> implements IApi<T> {
             pattern = new RegExp(`${pattern}$`);
         }
 
-        const newPlugins = this.plugins.slice();
-        newPlugins.push({ pattern, plugin } as IPluginEntry<T & P>);
-        return new Documentalist(newPlugins as IPluginEntry<T & P>[], this.markedOptions);
+        const newPlugins = [...this.plugins, { pattern, plugin } as IPluginEntry<T & P>];
+        return new Documentalist(newPlugins, this.markedOptions);
```
Comment (themadcreator) [inline src/documentalist.ts]: how come this didn't work for me when i tried it? strange

File: src/documentalist.ts [hunk 3]
```diff
@@ -140,24 +140,23 @@ export class Documentalist<T> implements IApi<T> {
...
     }
 
     public clearPlugins(): IApi<void> {
-        return new Documentalist<{}>([], this.markedOptions);
+        return new Documentalist<void>([], this.markedOptions);
```
Comment (themadcreator) [inline src/documentalist.ts]: yeah, i switched to {} when trying the `T extends object` but `void` is better

File: src/documentalist.ts [hunk 4]
```diff
@@ -140,24 +140,23 @@ export class Documentalist<T> implements IApi<T> {
...
     }
 
-    public documentGlobs(...filesGlobs: string[]) {
+    public async documentGlobs(...filesGlobs: string[]) {
```
Comment (themadcreator) [inline src/documentalist.ts]: what's this devilry?

File: src/documentalist.ts (hunk 2 above)
Comment (giladgray) [inline src/documentalist.ts]: i'll bet it's from `splice`. TS can be smart about spreads.

File: src/documentalist.ts (hunk 3 above)
Comment (giladgray) [inline src/documentalist.ts]: compiler made me do it! 🔫

**Pull request merged**